drafts/str4d-stratum/draft1.rst

```diff
@@ -0,0 +1,511 @@
+- Client sends ``mining.authorize`` for their worker(s).
+- Server replies with the result of authorization.
+- Server sends ``mining.set_target``.
+- Server sends ``mining.notify`` with a new job.
+- Client mines on that job.
+- Client sends ``mining.submit`` for each solution found.
+- Server replies with whether the solution was accepted.
+- Server sends ``mining.notify`` again when there is a new job.
+
+Nonce Parts
+~~~~~~~~~~~
+
+In Bitcoin, blocks contain two nonces: the 4-byte block header nonce, and an
+extra nonce in the coinbase transaction [Bitcoin-Block]_. The original Stratum
+protocol splits this extra nonce into two parts: one set by the server (used
+for splitting the search space amongst connected miners), and the other iterated
+by the miner [Slushpool-Stratum]_. The nonce in Zcash's block header is 32 bytes
+long [Zcash-Block]_, and thus can serve both purposes simultaneously.
+
+We define two nonce parts:
+
+``NONCE_1``
+  The server MUST pick such that ``len(NONCE_1) < 32`` in bytes.
+
+``NONCE_2``
+  The miner MUST pick such that ``len(NONCE_2) = 32 - len(NONCE_1)`` in bytes.
+
+  In hex, ``lenHex(NONCE_2) = 64 - lenHex(NONCE_1)``, and both lengths are even.
+
+The nonce in the block header is the concatenation of ``NONCE_1`` and
+``NONCE_2`` in hex. This means that miner using bignum representations of nonce
+MUST increment by ``1 << len(NONCE_1)`` to avoid altering ``NONCE_1`` (because
+the encoding of nonce in the block header is little endian, in line with the
+other 32-byte fields [Bitcoin-Block]_ [Zcash-Block]_).
+
+.. [Zcash-Block] Daira Hopwood, Sean Bowe, Taylor Hornby, Nathan Wilcox.
+  "Block Headers". In: *Zcash Protocol Specification*.
+  Version 2016.0-beta-1.5, Section 6.3. September 22, 2016.
+  URL: https://github.com/zcash/zips/blob/master/protocol/protocol.pdf
+  (visited on 2016-09-24).
+
+Session Resuming
+~~~~~~~~~~~~~~~~
+
+Servers that support session resuming identify this by setting a ``SESSION_ID``
+in their initial response. Servers MAY set ``SESSION_ID`` to ``null`` to
+indicate that they do not support session resuming. Servers that do not set
+``SESSION_ID`` to ``null`` MUST cache the following information:
+
+- The session ID.
+- ``NONCE_1``
+- Any active job IDs.
+
+Servers MAY drop entries from the cache on their own schedule.
+
+When a miner connects using a previous ``SESSION_ID``:
+
+- If the cache contains the ``SESSION_ID``, the server's initial response MUST
+  be constructed from the cached information.
+
+- If the server does not recognise the session, the ``SESSION_ID`` in the
+  server's initial response MUST NOT equal the ``SESSION_ID`` provided by the
+  miner.
+
+Miners MUST re-authorize all workers upon resuming a session.
+
+Methods
+~~~~~~~
+
+``mining.subscribe()``
+----------------------
+
+Request:
+
+    {"id": 1, "method": "mining.subscribe", "params": ["*CONNECT_HOST*", *CONNECT_PORT*, "*MINER_USER_AGENT*", "*SESSION_ID*"]} ``\n``
```

Comment:
Note that my opinion here is different to for `mining.notify` and `mining.submit`, because here the semantics of existing parameters are identical to Bitcoin Stratum.